r-package/grf/tests/testthat/test_causal_forest.R

```diff
@@ -125,12 +125,105 @@ test_that("local linear causal forests with large lambda are equivalent to causa
    p = 6
    X = matrix(rnorm(n*p), n, p)
    W = rbinom(n, 1, 0.5)
-   TAU = 2*X[,1] + X[,2]
-   Y = W * TAU + rnorm(n)
+   tau= 2*X[,1] + X[,2]
+   Y = W * tau + rnorm(n)
 
    forest = causal_forest(X, Y, W, num.trees = 400)
    preds.ll = predict(forest, X, linear.correction.variables = 1:ncol(X), ll.lambda = 1e5)$predictions
    preds.cf = predict(forest)$predictions
 
    expect_true(mean((preds.ll - preds.cf)^2) < 0.02)
 })
+
+
+test_that("predictions are invariant to scaling of the sample weights.", {
+   n = 100
+   p = 6
+   X = matrix(rnorm(n*p), n, p)
+   e = 1/(1+exp(-2*X[,1]+3*X[,2]))
+   W = rbinom(n, 1, e)
+   tau = 2*X[,1] + X[,2]
+   Y = W * tau + rnorm(n)
+   e.cc = 1/(1+exp(-2*X[,1]))
+   sample.weights = 1/e.cc
+
+   forest.1 = causal_forest(X, Y, W, sample.weights = sample.weights, seed=1)
+   forest.2 = causal_forest(X, Y, W, sample.weights = 40*sample.weights, seed=1)
+   expect_true(all(abs(forest.2$predictions - forest.1$predictions) < 1e-10))
+})
+
+test_that("IPCC weighting in the training of a causal forest with missing data improves its complete-data MSE.", {
+   old.seed = .Random.seed
+
+   seed = 1
+   set.seed(seed)
+   n = 400
+   p = 6
+   e =
+   X = matrix(rnorm(n*p), n, p)
+   e = 1/(1+exp(-2*X[,1]+3*X[,2]))
+   W = rbinom(n, 1, e)
+   tau = 2*X[,1] + X[,2]
+   Y = W * tau + rnorm(n)
+
+   e.cc = 1/(1+exp(-2*X[,1]))
+   cc = as.logical(rbinom(n, 1, e.cc))
+   sample.weights = 1/e.cc
+
+   num.trees = 400
+   forest = causal_forest(X[cc,], Y[cc], W[cc], num.trees=num.trees, seed = seed)
+   boosted.forest = causal_forest(X[cc,], Y[cc], W[cc], orthog.boosting = TRUE, num.trees = num.trees, seed = seed)
+   weighted.forest = causal_forest(X[cc,], Y[cc], W[cc], sample.weights = sample.weights[cc], num.trees = num.trees, seed = seed)
+   boosted.weighted.forest = causal_forest(X[cc,], Y[cc], W[cc], sample.weights = sample.weights[cc], orthog.boosting = TRUE, num.trees = num.trees, seed = seed)
+   mse = function(f) { sum((predict(f,X) - tau)^2) }
```

Comment:
For non-censored observations, these predictions aren't out-of-bag (so their MSE may not reflect the actual accuracy of the forest).

Can we create a separate test set here, and do clean MSE evaluation on test set predictions, so there's no risk of weirdness?